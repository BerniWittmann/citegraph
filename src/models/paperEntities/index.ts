import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
