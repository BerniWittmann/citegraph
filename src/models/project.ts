export default class Project {
  name: string;
  id?: number;

  constructor ({ name, id }: Project) {
    this.id = id
    this.name = name
  }
}
