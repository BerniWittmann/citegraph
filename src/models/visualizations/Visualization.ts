import Transformer from '@/transformers/Transformer'

export interface TimePeriod {
  from: number
  to: number
}

export interface VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods?: Array<TimePeriod>
}

export interface VisualizationBaseFieldsDTO {
  id?: string;
  progress?: number;
  name: string;
  key: string;
  // eslint-disable-next-line camelcase
  time_periods?: Array<TimePeriod>
}

export default abstract class Visualization implements VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods: Array<TimePeriod>
  abstract parameters: Object;
  abstract data?: Object;
  abstract readonly key: string;
  static readonly key: string;
  static readonly title: string;
  static readonly imageUrl: string;
  static readonly description: string;
  static readonly longDescription: string;
  static readonly transformer: typeof Transformer;

  protected constructor ({ id, progress, name, timePeriods }: VisualizationBaseFields) {
    this.id = id
    this.progress = progress
    this.name = name
    this.timePeriods = timePeriods || []
  }
}
