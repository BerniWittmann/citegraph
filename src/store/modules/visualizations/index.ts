import { Module } from 'vuex'
import { VisualizationsState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const visualizations: Module<VisualizationsState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
