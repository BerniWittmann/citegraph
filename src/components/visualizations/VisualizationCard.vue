<template>
  <v-card :hover="true"
          class="visualization-card"
          height="100%">
    <div class="d-flex flex-no-wrap justify-space-between">
      <v-avatar
        class="ma-3"
        size="125"
        tile
      >
        <v-img contain :src="imageUrl"/>
      </v-avatar>
      <div class="flex-grow-1 d-flex flex-column align-center">
        <v-card-title>{{ visualization.name }}</v-card-title>
        <v-card-text class="progress-information text-center">
          <span v-if="isLoading">{{ $t('visualizations.card.processing', { percent: progress }) }}</span>
          <span v-else>{{ $t('visualizations.card.ready') }}</span>
        </v-card-text>
        <v-divider style="width: 90%" class="mx-auto"></v-divider>
        <v-card-actions>
          <v-btn color="primary" text>
            {{ $t('visualizations.card.show') }}
          </v-btn>
          <v-btn color="primary" text>
            {{ $t('visualizations.card.edit') }}
          </v-btn>
        </v-card-actions>
      </div>
    </div>
    <v-progress-linear color="#F37F21" background-color="#FBBF35" background-opacity="0.5" v-show="isLoading" class="progress" :value="progress" :max="100"></v-progress-linear>
  </v-card>
</template>
<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import Visualization from '../../models/visualizations/Visualization'
import { visualizationsKeyMap } from '@/models/visualizations'

@Component
export default class VisualizationCard extends Vue {
  @Prop() readonly visualization!: Visualization

  get visualizationClass () {
    return visualizationsKeyMap[this.visualization.key]
  }

  get imageUrl (): string {
    return `${process.env.BASE_URL}${this.visualizationClass.imageUrl}`
  }

  get progress (): number {
    return this.visualization.progress ? this.visualization.progress * 100 : 0
  }

  get isLoading (): boolean {
    return !!this.visualization.progress && (this.progress < 100)
  }
}
</script>

<style scoped lang="scss">
.progress-information {
  color: #F37F21;
}
</style>
