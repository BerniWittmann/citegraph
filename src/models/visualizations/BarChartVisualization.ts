import Visualization, { VisualizationBaseFields } from './Visualization'

export interface BarChartVisualizationFields {}

export default class BarChartVisualization extends Visualization implements BarChartVisualizationFields {
  static key: string = 'barchart'
  key: string = 'barchart'
  static title: string = 'visualizations.information.bar_chart.title'
  static description: string = 'visualizations.information.bar_chart.description'
  static longDescription: string = 'visualizations.information.bar_chart.long_description'
  static imageUrl: string = 'visualization_images/bar_chart.png'

  constructor ({ id, name, progress }: VisualizationBaseFields & BarChartVisualizationFields) {
    super({ id, name, progress })
  }
}
