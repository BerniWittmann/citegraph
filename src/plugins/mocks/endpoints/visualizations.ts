import { AxiosRequestConfig } from 'axios'
import MockAdapter from 'axios-mock-adapter'
import db from '../db'
import Project from '@/models/project'
import Visualization from '@/models/visualizations/Visualization'

const projects = db.getCollection('projects')
const visualizations = db.getCollection('visualizations')

function getProjectId (config: AxiosRequestConfig): number {
  const urlParts: Array<string> = config.url!.split('/')
  return parseInt(urlParts[4])
}

function getVisualizationId (config: AxiosRequestConfig): string {
  const urlParts: Array<string> = config.url!.split('/')
  return urlParts[6]
}

export function setupMocks (mock: MockAdapter) {
  mock.onGet(/\/projects\/\d+\/visualizations$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    return [200, visualizations.find()]
  })

  mock.onGet(/\/projects\/\d+\/visualizations\/\d+$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    const visualization: Resultset<Visualization> = visualizations.findOne({ 'id': getVisualizationId(config) })
    if (!visualization) {
      return [404]
    }
    return [200, visualization]
  })
}
