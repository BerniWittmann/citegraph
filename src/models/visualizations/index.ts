import WordCloudVisualization, { WordCloudVisualizationFields } from './WordCloudVisualization'
import { VisualizationBaseFields } from '@/models/visualizations/Visualization'
import BarChartVisualization, { BarChartVisualizationFields } from '@/models/visualizations/BarChartVisualization'

export const visualizations = [WordCloudVisualization, BarChartVisualization]

const visualizationsKeyMap: Record<string, any> = {}
visualizations.forEach((visualization) => {
  visualizationsKeyMap[visualization.key] = visualization
})

export {
  visualizationsKeyMap
}

export type VisalizationFields = VisualizationBaseFields & (
  WordCloudVisualizationFields &
  BarChartVisualizationFields
)
