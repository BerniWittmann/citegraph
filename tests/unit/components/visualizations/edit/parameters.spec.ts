import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../setupPlugins'

import VisualizationEditParametersComponent from '@/components/visualizations/edit/Parameters.vue'
import Visualization from '@/models/visualizations/Visualization'
import { visualizations } from '@/models/visualizations'
import BarChartVisualization from '@/models/visualizations/BarChartVisualization'
import BarChartParameters from '@/components/visualizations/edit/parameters/BarChartParameters.vue'

jest.useFakeTimers()

describe('components/visualizations/edit/Parameters.vue', () => {
  const getWrapper = (vis: Visualization = new BarChartVisualization({ name: 'Test' })) => {
    return shallowMount(VisualizationEditParametersComponent, {
      i18n,
      mocks: {
        $t: (key: string) => key
      },
      propsData: {
        visualization: vis
      }
    })
  }

  it('renders', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders if no correct visualization type was found', () => {
    const wrapper = getWrapper({ name: 'False', key: 'invalid', parameters: {}, timePeriods: [] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  visualizations.forEach(VisualizationClass => {
    it('renders the correct component for ' + VisualizationClass.key, () => {
      const wrapper = getWrapper(new VisualizationClass({ name: 'Visualization' }))
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('can emit the next step event', () => {
    const wrapper = getWrapper()
    const comp = wrapper.find(BarChartParameters)
    expect(comp.exists()).toBeTruthy()
    expect(wrapper.emitted('next-step')).toBeFalsy()
    comp.vm.$emit('submit')
    expect(wrapper.emitted('next-step')).toBeTruthy()
  })

  it('can emit the previous step event', () => {
    const wrapper = getWrapper()
    const comp = wrapper.find(BarChartParameters)
    expect(comp.exists()).toBeTruthy()
    expect(wrapper.emitted('previous-step')).toBeFalsy()
    comp.vm.$emit('back')
    expect(wrapper.emitted('previous-step')).toBeTruthy()
  })
})
