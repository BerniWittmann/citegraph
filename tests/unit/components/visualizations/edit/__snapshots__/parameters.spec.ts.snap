// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/visualizations/edit/Parameters.vue renders 1`] = `
<div class="visualization-add-parameters">
  <h1>visualizations.add.parameters.title</h1>
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <v-col-stub tag="div">
        <barchartparameters-stub visualization="[object Object]"></barchartparameters-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;

exports[`components/visualizations/edit/Parameters.vue renders if no correct visualization type was found 1`] = `
<div class="visualization-add-parameters">
  <h1>visualizations.add.parameters.title</h1>
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <v-col-stub tag="div">
        <!---->
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;

exports[`components/visualizations/edit/Parameters.vue renders the correct component for barchart 1`] = `
<div class="visualization-add-parameters">
  <h1>visualizations.add.parameters.title</h1>
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <v-col-stub tag="div">
        <barchartparameters-stub visualization="[object Object]"></barchartparameters-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;

exports[`components/visualizations/edit/Parameters.vue renders the correct component for wordcloud 1`] = `
<div class="visualization-add-parameters">
  <h1>visualizations.add.parameters.title</h1>
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <v-col-stub tag="div">
        <wordcloudparameters-stub visualization="[object Object]"></wordcloudparameters-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
